lib/Scheduler/bb_spill.cpp

```diff
@@ -899,32 +1002,131 @@ void BBWithSpill::SetupForSchdulng_() {
 }
 /*****************************************************************************/
 
-bool BBWithSpill::ChkCostFsblty(InstCount trgtLngth, EnumTreeNode *node) {
+bool BBWithSpill::ChkCostFsblty(InstCount trgtLngth, EnumTreeNode *node,
+                                InstCount &RPCost) {
+  bool fsbl = true;
+
+  if (isTwoPassEnabled()) {
+    if (!IsSecondPass())
+      fsbl = ChkCostFsbltyFrstPss(trgtLngth, node, RPCost);
+    else
+      fsbl = ChkCostFsbltyScndPss(trgtLngth, node, RPCost);
+  }
+
+  else
+    fsbl = ChkCostFsbltyWghtd(trgtLngth, node);
+
+  return fsbl;
+}
+
+/*****************************************************************************/
+
+bool BBWithSpill::ChkCostFsbltyFrstPss(InstCount trgtLngth, EnumTreeNode *node,
+                                       InstCount &RPCost) {
+  bool fsbl;
+  InstCount TmpSpillCost;
+
+  InstCount crntCost;
+  if (GetSpillCostFunc() == SCF_SLIL) {
+    TmpSpillCost = dynamicSlilLowerBound_;
+    // this line currently needed for hist dom
+    crntCost = dynamicSlilLowerBound_ * SCW_ + trgtLngth * schedCostFactor_;
+    fsbl = dynamicSlilLowerBound_ < getBestSpillCost();
+  }
+
+  else {
+    TmpSpillCost = crntSpillCost_;
+    crntCost = crntSpillCost_ * SCW_ + trgtLngth * schedCostFactor_;
+    fsbl = crntSpillCost_ < getBestSpillCost();
+  }
+
+  crntCost -= GetCostLwrBound();
+
+  // assert(cost >= 0);
+  assert(crntCost >= 0);
+
+  if (fsbl) {
+    node->SetCost(crntCost);
+    node->SetCostLwrBound(crntCost);
+    node->SetPeakSpillCost(peakSpillCost_);
+    node->SetSpillCostSum(totSpillCost_);
+    node->setSpillCost(TmpSpillCost);
+    node->setSpillCostLwrBound(TmpSpillCost);
+  } else
+    RPCost = TmpSpillCost;
+
+  return fsbl;
+}
+
+/*****************************************************************************/
+
+bool BBWithSpill::ChkCostFsbltyScndPss(InstCount trgtLngth, EnumTreeNode *node,
+                                       InstCount &RPCost) {
+  InstCount TmpSpillCost;
+  InstCount crntCost;
+
+  if (GetSpillCostFunc() == SCF_SLIL) {
+    crntCost = dynamicSlilLowerBound_ * SCW_ + trgtLngth * schedCostFactor_;
+    TmpSpillCost = dynamicSlilLowerBound_;
+  }
+
+  else {
+    crntCost = crntSpillCost_ * SCW_ + trgtLngth * schedCostFactor_;
+    TmpSpillCost = crntSpillCost_;
+  }
+
+  crntCost -= GetCostLwrBound();
+
+  // assert(cost >= 0);
+  assert(crntCost >= 0);
+
+  if (TmpSpillCost <= getSpillCostConstraint()) {
+    node->SetCost(crntCost);
+    node->SetCostLwrBound(crntCost);
+    node->SetPeakSpillCost(peakSpillCost_);
+    node->SetSpillCostSum(totSpillCost_);
+    node->setSpillCost(TmpSpillCost);
+    node->setSpillCostLwrBound(TmpSpillCost);
+    return true;
+  }
+
+  RPCost = TmpSpillCost;
+  return false;
+}
+
+/*****************************************************************************/
+
+bool BBWithSpill::ChkCostFsbltyWghtd(InstCount trgtLngth, EnumTreeNode *node) {
```

Comment:
Similarly, there's duplicated code between these 3 functions. There has to be some way to reduce the duplication without cluttering it with a bunch of conditionals.